### PR TITLE
Define compdef.Shutdowner as a non-fx interface for app shutdown

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

Define a `compdef` version of `Shutdowner`, with an adapter in `fxutil`. This will enable more components to switch to the new folder layout, as before they were unable to remove fx from their implementation.

After this PR, components can use `compdef.Shutdowner` in their impl folder, and the `ProvideComponentConstructor` helper will adapt this interface to the actual `fx.Shutdowner`.

### Motivation

Multiple components use fx.Shutdowner and have been unable to convert fully to the new folder layout.

### Describe how to test/QA your changes

Functionality covered by unit tests.

### CODE REVIEW

File: comp/def/shutdowner.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package compdef defines basic types used for components
+package compdef
+
+// Shutdowner may be added to a component's requires struct so it can shutdown the application
```
Comment (GustavoCaso) [inline comp/def/shutdowner.go:9]:
```suggestion
// Shutdowner may be added to a component's dependencies so it can shutdown the application
```

Comment (GustavoCaso) [reply comp/def/shutdowner.go:9]: It seems that mentioning the require struct might be to low level for a package description. Let me know what you think

File: pkg/util/fxutil/provide_comp.go [hunk 2]
```diff
@@ -293,13 +294,23 @@ func coerceStructTo(input reflect.Value, outType reflect.Type, oldEmbed, newEmbe
 	return result
 }
 
+// FxPrerequisites returns all of our adapters from compdef types to fx types
+func FxPrerequisites() fx.Option {
```
Comment (GustavoCaso) [inline pkg/util/fxutil/provide_comp.go:298]: What about renaming to something that feels more custom to the Agent? 

`FxAgentOptions` or `AgentFxOptions` or `CustomFxAgentOptions`

What do you think?

File: pkg/util/fxutil/provide_comp_test.go [hunk 3]
```diff
@@ -451,6 +453,46 @@ func TestFxReturnAnError(t *testing.T) {
...
+	// create a wait-loop that will timeout if it takes longer than 1 second
+	ctx, cancel := context.WithCancel(context.Background())
+	time.AfterFunc(1*time.Second, cancel)
+	select {
+	case <-ctx.Done():
+		assert.Fail(t, "waiting for Application timed out, Shutdown() did not work")
+		return
+	case <-fxApp.Done():
+		// if compdef.Shutdown is successfully upgraded to fx.Shutdown,
+		// then this loop should exit here
```
Comment (GustavoCaso) [inline pkg/util/fxutil/provide_comp_test.go:489]: What about adding a OnStop lifecyle hook to the test app that when called it flip a boolean value. Then we can also assert that the value is true. 

I'm not sure if that is a bit redundant. What do you think?

**Review by GustavoCaso — APPROVED** [review]
LGTM. Left a few nitpick comments 😄

File: pkg/util/fxutil/provide_comp.go (hunk 2 above)
Comment (dustmop) [reply pkg/util/fxutil/provide_comp.go:298]: Hmmm, that's a fine suggestion. These `fx.Option`s being added are necessary for a lot of the fxutil wrappers, they assume that certain types are being provided, and I would like to capture that requirement in the name. How about `FxAgentBase()`?

File: pkg/util/fxutil/provide_comp_test.go (hunk 3 above)
Comment (dustmop) [reply pkg/util/fxutil/provide_comp_test.go:489]: Tried that approach, but I found that it 1) doesn't work as well and 2) complicates the logic of the test a lot. I like how this test more directly demonstrates exactly what `Shutdown` does.

**Review by daniel-taf — APPROVED** [review]

**Pull request merged**